### ISSUE TITLE
Incorrect values written to `pmpaddr` CSRs

### ISSUE THREAD

**Issue by udinator** [riscv-admin/riscv-ovpsim#15]
Hi,

This is a follow-up issue to Lee's comment in the Ibex repo.

I'm seeing some odd behavior in OVPsim after writes to any `pmpaddr` CSR - the situation is as follows.
![Screenshot from 2020-08-06 13-47-44](https://user-images.githubusercontent.com/16736281/89581217-721cd500-d7eb-11ea-8787-41f4831e97f7.png)
The above snippet of OVPsim log shows assembly code that writes a randomly generated value into `pmpaddr0` and then restores the original value afterwards.

The random value to be written is generated by the `add` instruction, `0xa3442a17`.
We can see that when we write `s3` into `pmpaddr0`, somehow the top bit gets masked off, so the value that actually ends up getting written into `pmpaddr0` is `0x23442a17`.
The assembly code then continues on to do the same thing with every other `pmpaddr` CSR, and in all cases a similar sort of mismatch results.

However, core Ibex (as well as Spike ISS) write the expected value of `0xa3442a17` into `pmpaddr0`, so the resultant mismatch when comparing Ibex logs with OVPsim logs causes failures in all of these PMP tests.

For context, all 16 PMP regions are utilized, with a granularity of 0 - the 8-bit configuration field for every single region is set to `0x0f`.

To reproduce the error with VCS, follow these steps:

```console
Prerequisites: OVPSIM_PATH and RISCV_TOOLCHAIN environment variables need to be set

git clone git@github.com:lowRISC/ibex.git

cd ibex/dv/uvm/core_ibex

make SEED=25882 ITERATIONS=1 TEST=riscv_pmp_basic_test
```
This test will fail.

The OVPsim log will be located at `out/seed-25882/instr_gen/ovpsim_sim/riscv_pmp_basic_test.0.log`, and the Ibex trace log will be located at `out/seed-25882/rtl_sim/riscv_pmp_basic_test.0/trace_core_00000000.log`.
The inconsistent behavior described above will be located at the same address as in the image, at `PC=0x8000025a`.
The OVPsim configuration file that we use for the Ibex DV environment is found at `riscv_dv_extension/riscvOVPsim.ic`.

Could you please provide some guidance on how to resolve this issue?
Thanks!

**Comment by Imperas** [comment]
@udinator 
Thanks for the description. We were able to create a small example showing the behavior you describe.
The behavior you are seeing is correct, for the setup you have defined. To see the desired behavior, you need to change one of the default settings - full description follows-

The default physical memory size in riscvOVPsim is 32 bits (for the processor and memory)
The pmpaddr register is word aligned, and contains bits 33:2 of the address
From the priv spec, 

> Each PMP address register encodes bits 33–2 of a 34-bit physical address for RV32 ...
> **Not all physical address bits may be implemented, and so the pmpaddr registers are WARL**.

So what you are seeing is correct behavior for a processor capable of addressing a physical memory size of 32 bits. I **think** what you are looking to do, is to define a physical address range of 34 bits, in which case simply tell riscvOVPsim that the addressbits should be 34 (not the default 32)
Add the following line to riscv_dv_extension/riscvOVPsim.ic.
--addressbits 34

I tested this writing a simple application based upon your assembler description, and it produced the same behavior you describe when set to 32 bits (default) and the desired behavior when set to 34 bits.

Thx
Lee

**Comment by udinator** [comment]
Hi Lee!

Ah, I see what is going on based on what you were saying, seems like since the default memory size is 32 bits, the top two bits of each `pmpaddr` CSR are masked off.
I wasn't aware of the `--addressbits` OVPsim setting, thank you for mentioning this! 
Using a 34-bit physical memory size seems to fix the issues going on with the simulation.

Thanks for the help!

**Comment by Imperas** [comment]
@udinator 
great thanks for the feedback, closing the issue
